### DISCUSSION TITLE
One-to-many relation from multiple models to the same model

### DISCUSSION THREAD

**Discussion by ugultopu** [prisma/prisma#21488] [Q&A]
### Question

Imagine that a user can have multiple images. Likewise, a product can have multiple images. However, a given image can belong to only one user or product.

In an RDBMS, I believe this would be modeled with `user_image` and `product_image` tables, where these tables having a FK to the `user` and `product` tables respectively.

I would like to achieve the same thing with Prisma, but I don't want any code duplication for better maintainability. That is, I don't want to have `user_image` and `product_image` models with the exact same attributes, except for the FK attributes. I would like the ability to define the common attributes in a separate model (such as `image`) and make `user_image` and `product_image` extend this model.

I guess one option is composition. That is, having an `image` model, a `user_image` model with a FK to the `image` model and another FK to the `user` model. However, from a data modeling perspective I believe this doesn't make much sense, since each entry in `image` will be referenced exactly once, either from `user_image` or from `product_image`. So, this doesn't seem right from a data modeling perspective.

What are my options?

**Comment by ludralph** [comment]
Hi @ugultopu 👋 

Based on your question, I created the schema below:

```prisma
model Image {
  id          Int       @id @default(autoincrement())
  url         String
  description String
  userImages UserImage[]
  productImages ProductImage[]
}

model UserImage {
  id       Int       @id @default(autoincrement())
  image   Image     @relation(fields: [imageId], references: [id])
  imageId Int
  user    User      @relation(fields: [userId], references: [id])
  userId  Int
}

model ProductImage {
  id       Int       @id @default(autoincrement())
  image   Image     @relation(fields: [imageId], references: [id])
  imageId Int
  product Product    @relation(fields: [productId], references: [id])
  productId Int
}

model User {
  id          Int           @id @default(autoincrement())
  email       String        @unique
  name        String?
  images      UserImage[]
}

model Product {
  id          Int           @id @default(autoincrement())
  name        String
  description String
  price       Float
  images      ProductImage[]
}
```

I'm using separate join tables (`UserImage` and `ProductImage`) that establish the connections between `User` or `Product` entities and the `Image` entity. This setup ensures that each image is uniquely associated with either a user or a product, fulfilling the one-to-many relationship requirement without code duplication.

Additionally, the inclusion of separate `UserImage` and `ProductImage` models allows for specific attributes or fields related to the relationship between users/products and images to be captured, enabling more customized information storage while maintaining the necessary associations.


If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
